odbc_fdw.c

```diff
@@ -1358,8 +1433,87 @@ odbcIterateForeignScan(ForeignScanState *node)
 			   SQL_C_TYPE_DATE/SQL_C_TYPE_TIME/SQL_C_TYPE_TIMESTAMP.
 			   And finally, SQL_C_NUMERIC and SQL_C_GUID could also be used.
 			*/
+			buf[0] = 0;
+			buf_used = 0;
 			ret = SQLGetData(stmt, i, SQL_C_CHAR,
 							 buf, sizeof(char) * (col_size+1), &indicator);
+			buf_used = indicator;
+
+			if (ret == SQL_SUCCESS_WITH_INFO)
+			{
+				SQLCHAR sqlstate[5];
+				SQLGetDiagRec(SQL_HANDLE_STMT, stmt, 1, sqlstate, NULL, NULL, 0, NULL);
+				if (strcmp((char*)sqlstate, "01S07") == 0)
+				{
+					/* Fractional truncation has occured;
+					 * at this point we cannot obtain the lost digits
+					 */
+					if (buf[col_size])
+					{
+						/* The driver has omitted the trailing */
+						char *buf2 = (char *) palloc(sizeof(char) * (col_size+2));
+						strncpy(buf2, buf, col_size+1);
+						buf2[col_size+1] = 0;
+						pfree(buf);
+						buf = buf2;
+					}
+					elog(NOTICE,"Truncating number: %s",buf);
+				}
+				else
+				{
+					/* The output is incomplete, we need to obtain the rest of the data */
+					char* accum_buffer;
+					size_t accum_buffer_size;
+					size_t accum_used;
+					if (indicator == SQL_NO_TOTAL)
+					{
+						/* Unknown total size, must copy part by part */
+						accum_buffer_size = 0;
+						accum_buffer = NULL;
+						accum_used = 0;
+						while (1)
+						{
+							size_t buf_len = buf[col_size] ? col_size + 1 : col_size;
+							// Allocate new accumulation buffer if necessary
+							if (accum_used + buf_len > accum_buffer_size)
+							{
+								char *new_buff;
+								accum_buffer_size = accum_buffer_size == 0 ? col_size*2 : accum_buffer_size*2;
+								new_buff = (char *) palloc(sizeof(char) * (accum_buffer_size+1));
+								if (accum_buffer)
+								{
+									memmove(new_buff, accum_buffer, accum_used);
+									pfree(accum_buffer);
+								}
+								accum_buffer = new_buff;
+								accum_buffer[accum_used] = 0;
+							}
+							// Copy part to the accumulation buffer
+							strncpy(accum_buffer+accum_used, buf, buf_len);
+							accum_used += buf_len;
+							accum_buffer[accum_used] = 0;
+							// Get new part
+							if (ret != SQL_SUCCESS_WITH_INFO)
+							  break;
```

Comment:
I'll give it a try at rewriting this.

Than middle-of-the-loop exit point must eventually be take because `GetData` only returns `SQL_SUCCESS_WITH_INFO` until the whole data has been read.